pkg/operator/admission/webhook.go

```diff
@@ -0,0 +1,170 @@
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"os"
+)
+
+const (
+	secretName           = "rook-admission-controller"
+	admissionServiceName = "rook-admission-controller"
+	secretVolumeName     = "webhook-certs"
+	secretMountPath      = "/etc/webhook"
+	serviceAccountName   = "rook-admission-controller-sa"
+	storageFlag          = "ceph"
+)
+
+var (
+	validatePath       = "/validate"
+	servicePort  int32 = 443
+	serverPort   int32 = 8079
+	namespace          = os.Getenv(k8sutil.PodNamespaceEnvVar)
+)
+
+func isSecretPresent(context *clusterd.Context) (bool, error) {
+
+	_, err := context.Clientset.CoreV1().Secrets(namespace).Get(secretName, metav1.GetOptions{})
+	if err != nil {
+		// If secret is not found. All good ! Proceed with rook without admission controllers
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func createWebhookService(context *clusterd.Context) error {
+
+	webhookService := corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      admissionServiceName,
+			Namespace: namespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Port: servicePort,
+					TargetPort: intstr.IntOrString{
+						IntVal: serverPort,
+					},
+				},
+			},
+
+			Selector: map[string]string{
+				"app": "rook-ceph-admission-controller",
+			},
+		},
+	}
+	_, err := context.Clientset.CoreV1().Services(namespace).Create(&webhookService)
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		return err
+	}
+	return nil
+}
+
+// InitWebhookIfSecretPresent will initialize the webhook if secret is detected
+func InitWebhookIfSecretPresent(context *clusterd.Context, admissionImage string) error {
+	isPresent, err := isSecretPresent(context)
+	if isPresent {
+		err = initWebhook(context, admissionImage)
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize webhook")
+		}
+	}
+	if err != nil {
+		return errors.Wrap(err, "failed to retrieve secret")
+	}
+
+	return nil
+}
+
+func initWebhook(context *clusterd.Context, admissionImage string) error {
+	// At this point volume should be mounted, so proceed with creating the service and validatingwebhookconfig
+	err := createWebhookService(context)
+	if err != nil {
+		return errors.Wrap(err, "failed to create service.")
+	}
+	createWebhookPods(context, admissionImage)
+
+	return nil
+}
+
+func createWebhookPods(context *clusterd.Context, admissionImage string) error {
+	admission_parameters := []string{"admission-controller",
+		"--storage", storageFlag}
+	secretVolume := corev1.Volume{
+		Name: secretVolumeName,
+		VolumeSource: corev1.VolumeSource{
+			Secret: &corev1.SecretVolumeSource{
+				SecretName: secretName,
+			},
+		},
+	}
+	secretVolumeMount := corev1.VolumeMount{
+		Name:      secretVolumeName,
+		ReadOnly:  true,
+		MountPath: secretMountPath,
+	}
+
+	var replicas int32 = 2
```

Comment:
Since there are two replicas, they should have pod anti-affinity to each other. See #5462 for an example of other pod anti-affinity recently added to the csi provisioner.

Reply:
I've added this.

Reply:
This should be resolved now. PodAntiAffinity is working as expected.